tests/50federation/40devicelists.pl

```diff
@@ -284,3 +284,132 @@
+            $outbound_client_user => {
+               CURIOSITY_ROVER => {
+                  user_id => $outbound_client_user,
+                  device_id => "CURIOSITY_ROVER",
+                  algorithms => ["fast", "and broken"],
+                  keys => {
+                     "curve25519:JLAFKJWSCS" => "3C5BFWi2Y8MaVvjM8M22DBmh24PmgR0nPvJOIArzgyI",
+                     "ed25519:JLAFKJWSCS" => "lEuiRJBit0IG6nUf5pUzWTUEsRVVe/HJkoKuEww9ULI"
+                  },
+                  signatures => {
+                     $outbound_client_user => {"ed25519:JLAFKJWSCS" => "dSO80A01XiigH3uBiDVx/EjzaoycHcjq9lfQX0uWsqxl2giMIiSPR8a4d291W1ihKJL/a+myXS367WT6NAIcBA"}
+                  },
+                  unsigned => {
+                     "device_display_name" => "Curiosity Rover"
+                  },
+               },
+            }
+         },
+      };
+
+      Future->needs_all(
+         $inbound_server->await_request_user_keys_query( $outbound_client_user )->then( sub {
+            my ( $req ) = @_;
+            $req->respond_json($client_user_keys);
+            Future->done(1)
+         }),
+
+         matrix_create_room(
+            $local_user,
+            preset => "public_chat",
+            room_version => "5",
+         )->then( sub {
+            my ( $room_id ) = @_;
+            $outbound_client->join_room(
+               server_name => $local_server_info->server_name,
+               room_id     => $room_id,
+               user_id     => $outbound_client_user,
+            )
+         })->then( sub {
+            do_request_json_for( $local_user,
+               method  => "POST",
+               uri     => "/r0/keys/query",
+               content => {
+                  device_keys => {
+                     $outbound_client_user => {}
+                  }
+               }
+            )
+         })->then( sub {
+            ( $first_keys_query_body ) = @_;
+            delete $first_keys_query_body->{failures};
+            assert_deeply_eq( $first_keys_query_body, $client_user_keys, "Local user gets federation user's keys."),
+            Future->done(1)
+         })
+      )->then( sub {
+         my $respond_503_fun = sub {
+            my ( $req ) = @_;
+            $OUTPUT->diag("responding");
+            $req->respond_json({}, code => 503);
+            Future->done(1)
+         };
+         my @respond_503 = (
+            $inbound_server->await_request_user_devices( $outbound_client_user )->then( sub {
+               my ( $req ) = @_;
+               $OUTPUT->diag("responding");
+               $req->respond_json({}, code => 503);
+               Future->done(1)
+            }),
+            $inbound_server->await_request_user_keys_query()->then( sub {
+               my ( $req ) = @_;
+               $OUTPUT->diag("responding");
+               $req->respond_json({}, code => 503);
+               Future->done(1)
+            })
+         );
+         Future->wait_any(
+            @respond_503,
```

Comment:
Won't `Future->wait_any` return here when one of the above `await_*` returns? I think you probably want to not wait on them and just return the `do_request_json`?

Reply:
Rookie question: if I do so how do I start the two futures I've created using await*?